Add networking support

mborgerson opened the pull request:
This patch adds basic networking capabilities to XQEMU. Currently the emulation is not exactly hardware-accurate, but is functional enough for basic packet tx/rx with both the official Microsoft Xbox networking libraries and NXDK networking libraries.

Once merged, networking _should_ just work out of the box. Here are some options a user might like to use as well:
- To forward a port (8080 on host to 21 on guest, for instance), add this option: `-redir tcp:8080::21`
- To dump a listing of all packets sent or received, which can then be interpreted by a tool such as Wireshark, add this option: `-global nvnet.dump=packets.txt`

Matt

CodeAsm commented:
Thats some serious great progress. I might try this. Im not entirely great with qemu but with redirecting i could link it to another (normal) qemu instance? Or is this a early stage to see basic networking with a workstation (your reseach pc)?

mborgerson commented:
@CodeAsm Thanks. You can allow another instance to connect to the VLAN by starting XQEMU with these options:

```
-net nic,model=nvnet -net user
-net socket,listen=:1236 
```

Then on the instance of QEMU you wish to join to the LAN something like this:

```
-net socket,connect=127.0.0.1:1236
```

espes commented:
Looks good! Just a bunch of style nits picked up by checkpatch.pl: https://gist.github.com/espes/b00da1d9868c1d8676f3, since you're adding a bunch of new code.

espes commented on hw/xbox/nvnet.c:
```diff
@@ -19,78 +20,983 @@
+{
+    struct RingDesc desc;
+
+    for (int i = 0; i < s->rx_ring_size; i++)
```
nit: not picked up by checkpatch, but no qemu code seems to use inline for initialisers for some reason

espes commented on hw/xbox/acpi_xbox.c:
```diff
@@ -33,7 +33,6 @@
 
 #include "hw/xbox/acpi_xbox.h"
 
-#define DEBUG
```
changes to this file needn't be in this PR

mborgerson committed:
Add networking support

This patch adds basic networking capabilities to XQEMU. Currently the
emulation is not exactly hardware-accurate, but is functional enough
for basic packet tx/rx with both the official Microsoft Xbox networking
libraries and NXDK networking libraries.

mborgerson commented:
@espes Thanks for the feedback. I was unaware of checkpatch.pl. I've incorporated all of your feedback: checkpatch.pl is reporting nothing now, I've moved those iterator var definitions out, and removed the unrelated modified file that snuck in to my commit. Let me know if there's anything else.

Matt

espes commented:
Thanks!
